src/Microsoft.AspNetCore.OData/EnableNestedPathsAttribute.cs

```diff
@@ -0,0 +1,120 @@
+        public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+        {
+            if (actionExecutedContext == null)
+            {
+                throw Error.ArgumentNull("actionExecutedContext");
+            }
+
+            HttpRequest request = actionExecutedContext.HttpContext.Request;
+            if (request == null)
+            {
+                throw Error.Argument("actionExecutedContext", SRResources.ActionExecutedContextMustHaveRequest);
+            }
+
+            ActionDescriptor actionDescriptor = actionExecutedContext.ActionDescriptor;
+            if (actionDescriptor == null)
+            {
+                throw Error.Argument("actionExecutedContext", SRResources.ActionContextMustHaveDescriptor);
+            }
+
+
+
+            //base.OnActionExecuted(context);
+            HttpResponse response = actionExecutedContext.HttpContext.Response;
+
+            ObjectResult responseContent = actionExecutedContext.Result as ObjectResult;
+            var result = responseContent.Value as IQueryable;
+            SingleResult singleResult = responseContent.Value as SingleResult;
+            if (singleResult != null)
+            {
+                // This could be a SingleResult, which has the property Queryable.
+                // But it could be a SingleResult() or SingleResult<T>. Sort by number of parameters
+                // on the property and get the one with the most parameters.
+                PropertyInfo propInfo = responseContent.Value.GetType().GetProperties()
+                    .OrderBy(p => p.GetIndexParameters().Count())
+                    .Where(p => p.Name.Equals("Queryable"))
+                    .LastOrDefault();
+
+                result = propInfo.GetValue(singleResult) as IQueryable;
+            }
+
+            
+            if (result != null)
+            {
+                ODataPath path = actionExecutedContext.HttpContext.ODataFeature().Path;
+                IEdmModel model = actionExecutedContext.HttpContext.Request.GetModel();
+
+                var queryBuilder = new ODataPathQueryBuilder(result, model, path);
+                ODataPathQueryResult transformedResult = queryBuilder.BuildQuery();
+
+                if (transformedResult == null)
+                {
+                    // TODO: is this the best way to return 404
+                    actionExecutedContext.Result = new NotFoundObjectResult(null);
+                }
+                else if (path.EdmType.TypeKind == EdmTypeKind.Collection || transformedResult.HasCountSegment)
+                {
+                    responseContent.Value = transformedResult.Result;
+                }
+                else
+                {
+                    Type elementType = transformedResult.Result.ElementType;
+                    object singleValue = EnableQueryAttribute.SingleOrDefault(transformedResult.Result, new WebApiActionDescriptor(actionDescriptor as ControllerActionDescriptor));
```

Comment:
@habbes Not sure I understand what is happening here...

Reply:
The result returned by `ODataPathQueryBuilder.BuildQuery()` is an IQueryable, if you return it as-is, then it will always return a collection response. So here I'm checking whether the result of the path is expected to be a single item or a collection, if it's a single item hen I call the `SingleOrDefault` helper method to extract the single item from the `IQueryable` and return it.

It's a helper method used by `EnableQueryAttribute`, but now I think I should extract it to a common helper class.

Reply:
I think a potential concern here is potentially materializing the IQueryable prematurely, if the controller action has both `[EnableQuery]` and `[EnableNestedPaths]`. If the `EnableNestedPathsAttribute` calls `SingleOrDefault` and sets it as result, I believe this materializes the IQueryable (i.e. executes the db query), now when this single object is later processed by `[EnableQuery]`, it will be applying the query options like select, expand to the single object in memory, as opposed to the deferred IQueryable. In the case of $select, this means that we would have fetched all the fields from the database, then doing the field selection on the object in memory. In the case of $expand, I'm not sure it would have fetched the necessary related entity, so I have to verify whether this will actually expand the expected property or just set it to null.

In this case, I think I should check whether the controller action also has `[EnableQuery]`, if it has it, then instead of calling `SingleOrDefault`, I create a `SingleResult` wrapper for the returned `IQueryable` then `[EnableQuery]` will do the rest, if the action does not have `[EnableQuery]` then I just call `SingleOrDefault`

Reply:
My suspicion was correct, using SingleOrDefault() to return a single result when the action also has `[EnableQuery]` results in query options like `$expand` returning null. If fixed this by using `SingleOrDefault()` if the action **does not** have `[EnableQuery]` and use `SingleResult<T>` if the action has `[EnableQuery]`